### PR TITLE
L0 flush: opt-in mechanism to bypass PageCache reads and writes

### PR DESCRIPTION
**Pull request by problame**
part of https://github.com/neondatabase/neon/issues/7418

# Motivation

(reproducing #7418)

When we do an `InMemoryLayer::write_to_disk`, there is a tremendous amount of random read I/O, as deltas from the ephemeral file (written in LSN order) are written out to the delta layer in key order.

In benchmarks (https://github.com/neondatabase/neon/pull/7409) we can see that this delta layer writing phase is substantially more expensive than the initial ingest of data, and that within the delta layer write a significant amount of the CPU time is spent traversing the page cache.

# High-Level Changes

Add a new mode for L0 flush that works as follows:

* Read the full ephemeral file into memory -- layers are much smaller than total memory, so this is afforable
* Do all the random reads directly from this in memory buffer instead of using blob IO/page cache/disk reads.
* Add a semaphore to limit how many timelines may concurrently do this (limit peak memory).
* Make the semaphore configurable via PS config.

# Implementation Details

The new `BlobReaderRef::Slice` is a temporary hack until we can ditch `blob_io` for `InMemoryLayer` => Plan for this is laid out in https://github.com/neondatabase/neon/issues/8183

# Correctness

The correctness of this change is quite obvious to me: we do what we did before (`blob_io`) but read from memory instead of going to disk.

The highest bug potential is in doing owned-buffers IO. I refactored the API a bit in preliminary PR https://github.com/neondatabase/neon/pull/8186 to make it less error-prone, but still, careful review is requested.

# Performance

I manually measured single-client ingest performance from `pgbench -i ...`.

Full report: https://neondatabase.notion.site/2024-06-28-benchmarking-l0-flush-performance-e98cff3807f94cb38f2054d8c818fe84?pvs=4

tl;dr:

* no speed improvements during ingest,  but
* significantly lower pressure on PS PageCache (eviction rate drops to 1/3)
  * (that's why I'm working on this)
* noticable but modestly lower CPU time

This is good enough for merging this PR because the changes require opt-in.

We'll do more testing in staging & pre-prod.

# Stability / Monitoring

**memory consumption**: there's no _hard_ limit on max `InMemoryLayer` size (aka "checkpoint distance") , hence there's no hard limit on the memory allocation we do for flushing. In practice, we a) [log a warning](https://github.com/neondatabase/neon/blob/23827c6b0d400cbb9a972d4d05d49834816c40d1/pageserver/src/tenant/timeline.rs#L5741-L5743) when we flush oversized layers, so we'd know which tenant is to blame and b) if we were to put a hard limit in place, we would have to decide what to do if there is an InMemoryLayer that exceeds the limit.
It seems like a better option to guarantee a max size for frozen layer, dependent on `checkpoint_distance`. Then limit concurrency based on that.

**metrics**: we do have the [flush_time_histo](https://github.com/neondatabase/neon/blob/23827c6b0d400cbb9a972d4d05d49834816c40d1/pageserver/src/tenant/timeline.rs#L3725-L3726), but that includes the wait time for the semaphore. We could add a separate metric for the time spent after acquiring the semaphore, so one can infer the wait time. Seems unnecessary at this point, though.

### CODE REVIEW

**Comment by problame** [comment]
the remote_storage changes will hopefully land before this PR in separate PR https://github.com/neondatabase/neon/pull/8193

File: pageserver/src/tenant/ephemeral_file/page_caching.rs [hunk 1]
```diff
@@ -49,6 +58,46 @@ impl RW {
         self.rw.bytes_written()
     }
 
+    /// Load all blocks that can be read via [`Self::read_blk`] into a contiguous memory buffer.
+    ///
+    /// This includes the blocks that aren't yet flushed to disk by the internal buffered writer.
+    /// The last block is zero-padded to [`PAGE_SZ`], so, the returned buffer is always a multiple of [`PAGE_SZ`].
+    pub(super) async fn load_into_contiguous_memory(
```
Comment (jcsp) [inline pageserver/src/tenant/ephemeral_file/page_caching.rs:65]: There's potential for confusion in this name: sometimes when an I/O function talks about contiguous memory it means physically contiguous.  Perhaps just `load_to_vec`?

File: pageserver/src/tenant/ephemeral_file/page_caching.rs [hunk 2]
```diff
@@ -49,6 +58,46 @@ impl RW {
...
+            } else {
+                s.checked_add(PAGE_SZ - (s % PAGE_SZ)).unwrap()
+            }
+        };
+        let buf = Vec::with_capacity(size);
+
+        // read from disk what we've already flushed
+        let writer = self.rw.as_writer();
+        let nwritten_blocks = usize::try_from(writer.nwritten_blocks).unwrap();
+        let buf = writer
```
Comment (jcsp) [inline pageserver/src/tenant/ephemeral_file/page_caching.rs:83]: Kind of weird to have the main read function (`load_flushed_blocks_into_contiguous_memory`) as part of the writer type.  Maybe just give the writer a function that returns the information needed to execute the read?

File: pageserver/src/tenant/ephemeral_file/page_caching.rs [hunk 3]
```diff
@@ -49,6 +58,46 @@ impl RW {
...
+                buf.slice(0..nwritten_blocks * PAGE_SZ),
+                ctx,
+            )
+            .await?;
+        let mut vec = buf.into_inner();
+        assert_eq!(vec.len() % PAGE_SZ, 0);
+        assert_eq!(vec.len() / PAGE_SZ, nwritten_blocks);
+
+        // copy from in-memory buffer what we haven't flushed yet but would return when accessed via read_blk
+        let buffered = self.rw.inspect_served_from_zero_padded_mutable_tail();
```
Comment (jcsp) [inline pageserver/src/tenant/ephemeral_file/page_caching.rs:94]: naming: maybe `get_tail_zero_padded`?  "inspect" sounds like something doing checks rather than something that's part of our I/O read path.

**Comment by problame** [comment]
Did some manual perf testing. Updated PR description, report here: https://www.notion.so/neondatabase/2024-06-28-benchmarking-l0-flush-performance-e98cff3807f94cb38f2054d8c818fe84?pvs=4

**Comment by problame** [comment]
> # Stability / Monitoring
> 
> TODO
> 

Existing page cache metrics are sufficient to quantify the impact on PageCache.

Do we want metrics on semaphore wait queue length or are higher-level metrics sufficient?

**Comment by problame** [comment]
@jcsp  I addressed your review comments, see latest pushes.

Also, given that I plumbed through the l0_flush::L0FlushGlobalState, maybe we want to move the `GlobalResourceUnits` there, so it no longer is a global lazy static?

https://github.com/neondatabase/neon/blob/511d6644c35b3b61e19da6766db7c194d6c481bd/pageserver/src/tenant/storage_layer/inmemory_layer.rs#L116-L121

**Review by jcsp — APPROVED** [review]
